cached_property is not ignored when a model is copied and updated

### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))


### Description

This is somehow a contrived situation, but we had it trigger in our business code. Basically when updating a model based on an existing one which has a subfield that is a dataclass with a cached_property, a validation is triggered because the dataclass `__dict__` contains the cached value. 

### Example Code

```Python
import dataclasses
from pydantic import BaseModel
from functools import cached_property

def test_cached_property():
    @dataclasses.dataclass(frozen=True)
    class A:
        name: str

        @cached_property
        def _name(self): 
            return "name"

    class MyModel(BaseModel, arbitrary_types_allowed=True, frozen=True, extra=Extra.forbid):
        scheduler: A

    models = {
        "AX": MyModel(scheduler=A("a")),
    }

    sched = A("sched")

    models_2 = {
        "AY": models["AX"].copy(update=dict(scheduler=sched)), 
    }

    models |= models_2

    models["AY"].scheduler._name
    MyModel.parse_obj(models["AY"].dict())
```

```
__pydantic_self__ = MyModel(), data = {'scheduler': test_cached_property.<locals>.A(name='sched')}, values = {}, fields_set = {'scheduler'}
validation_error = ValidationError(model='MyModel', errors=[{'loc': ('scheduler', '_name'), 'msg': 'extra fields not permitted', 'type': 'value_error.extra'}])

    def __init__(__pydantic_self__, **data: Any) -> None:
        """
        Create a new model by parsing and validating input data from keyword arguments.
    
        Raises ValidationError if the input data cannot be parsed to form a valid model.
        """
        # Uses something other than `self` the first arg to allow "self" as a settable attribute
        values, fields_set, validation_error = validate_model(__pydantic_self__.__class__, data)
        if validation_error:
>           raise validation_error
E           pydantic.error_wrappers.ValidationError: 1 validation error for MyModel
E           scheduler -> _name
E             extra fields not permitted (type=value_error.extra)

pydantic/main.py:341: ValidationError
```


### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.13
            pydantic compiled: False
                 install path: /home/rdb/git/pydantic/pydantic
               python version: 3.11.6 (main, Nov 14 2023, 09:36:21) [GCC 13.2.1 20230801]
                     platform: Linux-5.15.133.1-microsoft-standard-WSL2-x86_64-with-glibc2.38
     optional deps. installed: ['devtools', 'dotenv', 'email-validator', 'typing-extensions']
```


### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [X] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [X] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [X] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

--- a/pydantic/dataclasses.py
+++ b/pydantic/dataclasses.py
@@ -35,7 +35,7 @@ import copy
 import dataclasses
 import sys
 from contextlib import contextmanager
-from functools import wraps
+from functools import cached_property, wraps
 from typing import TYPE_CHECKING, Any, Callable, ClassVar, Dict, Generator, Optional, Type, TypeVar, Union, overload
 
 from typing_extensions import dataclass_transform
@@ -408,6 +408,8 @@ def create_pydantic_model_from_dataclass(
     model.__doc__ = dc_cls_doc if dc_cls_doc is not None else dc_cls.__doc__ or ''
     return model
 
+def _is_field_cached_property(obj: 'Dataclass', k: str) -> bool:
+    return isinstance(getattr(type(obj), k, None), cached_property)
 
 def _dataclass_validate_values(self: 'Dataclass') -> None:
     # validation errors can occur if this function is called twice on an already initialised dataclass.
@@ -417,9 +419,9 @@ def _dataclass_validate_values(self: 'Dataclass') -> None:
     if getattr(self, '__pydantic_has_field_info_default__', False):
         # We need to remove `FieldInfo` values since they are not valid as input
         # It's ok to do that because they are obviously the default values!
-        input_data = {k: v for k, v in self.__dict__.items() if not isinstance(v, FieldInfo)}
+        input_data = {k: v for k, v in self.__dict__.items() if not (isinstance(v, FieldInfo) or _is_field_cached_property(self, k))}
     else:
-        input_data = self.__dict__
+        input_data = {k: v for k, v in self.__dict__.items() if not _is_field_cached_property(self, k)}
     d, _, validation_error = validate_model(self.__pydantic_model__, input_data, cls=self.__class__)
     if validation_error:
         raise validation_error

--- a/pydantic/dataclasses.py
+++ b/pydantic/dataclasses.py
@@ -35,7 +35,14 @@ import copy
 import dataclasses
 import sys
 from contextlib import contextmanager
-from functools import cached_property, wraps
+from functools import wraps
+
+try:
+    from functools import cached_property
+except ImportError:
+    # cached_property available only for python3.8+
+    pass
+
 from typing import TYPE_CHECKING, Any, Callable, ClassVar, Dict, Generator, Optional, Type, TypeVar, Union, overload
 
 from typing_extensions import dataclass_transform
@@ -408,8 +415,17 @@ def create_pydantic_model_from_dataclass(
     model.__doc__ = dc_cls_doc if dc_cls_doc is not None else dc_cls.__doc__ or ''
     return model
 
-def _is_field_cached_property(obj: 'Dataclass', k: str) -> bool:
-    return isinstance(getattr(type(obj), k, None), cached_property)
+
+if sys.version_info >= (3, 8):
+
+    def _is_field_cached_property(obj: 'Dataclass', k: str) -> bool:
+        return isinstance(getattr(type(obj), k, None), cached_property)
+
+else:
+
+    def _is_field_cached_property(obj: 'Dataclass', k: str) -> bool:
+        return False
+
 
 def _dataclass_validate_values(self: 'Dataclass') -> None:
     # validation errors can occur if this function is called twice on an already initialised dataclass.
@@ -419,7 +435,11 @@ def _dataclass_validate_values(self: 'Dataclass') -> None:
     if getattr(self, '__pydantic_has_field_info_default__', False):
         # We need to remove `FieldInfo` values since they are not valid as input
         # It's ok to do that because they are obviously the default values!
-        input_data = {k: v for k, v in self.__dict__.items() if not (isinstance(v, FieldInfo) or _is_field_cached_property(self, k))}
+        input_data = {
+            k: v
+            for k, v in self.__dict__.items()
+            if not (isinstance(v, FieldInfo) or _is_field_cached_property(self, k))
+        }
     else:
         input_data = {k: v for k, v in self.__dict__.items() if not _is_field_cached_property(self, k)}
     d, _, validation_error = validate_model(self.__pydantic_model__, input_data, cls=self.__class__)

--- a/pydantic/error_wrappers.py
+++ b/pydantic/error_wrappers.py
@@ -101,7 +101,6 @@ def flatten_errors(
 ) -> Generator['ErrorDict', None, None]:
     for error in errors:
         if isinstance(error, ErrorWrapper):
-
             if loc:
                 error_loc = loc + error.loc_tuple()
             else:

--- a/pydantic/fields.py
+++ b/pydantic/fields.py
@@ -178,7 +178,6 @@ class FieldInfo(Representation):
         self.extra = kwargs
 
     def __repr_args__(self) -> 'ReprArgs':
-
         field_defaults_to_hide: Dict[str, Any] = {
             'repr': True,
             **self.__field_constraints__,
@@ -405,7 +404,6 @@ class ModelField(Representation):
         alias: Optional[str] = None,
         field_info: Optional[FieldInfo] = None,
     ) -> None:
-
         self.name: str = name
         self.has_alias: bool = alias is not None
         self.alias: str = alias if alias is not None else name
@@ -852,7 +850,6 @@ class ModelField(Representation):
     def validate(
         self, v: Any, values: Dict[str, Any], *, loc: 'LocStr', cls: Optional['ModelOrDc'] = None
     ) -> 'ValidateReturn':
-
         assert self.type_.__class__ is not DeferredType
 
         if self.type_.__class__ is ForwardRef:

--- a/pydantic/main.py
+++ b/pydantic/main.py
@@ -737,7 +737,6 @@ class BaseModel(Representation, metaclass=ModelMetaclass):
         exclude_defaults: bool,
         exclude_none: bool,
     ) -> Any:
-
         if isinstance(v, BaseModel):
             if to_dict:
                 v_dict = v.dict(
@@ -830,7 +829,6 @@ class BaseModel(Representation, metaclass=ModelMetaclass):
         exclude_defaults: bool = False,
         exclude_none: bool = False,
     ) -> 'TupleGenerator':
-
         # Merge field set excludes with explicit exclude parameter with explicit overriding field set options.
         # The extra "is not None" guards are not logically necessary but optimizes performance for the simple case.
         if exclude is not None or self.__exclude_fields__ is not None:

--- a/pydantic/schema.py
+++ b/pydantic/schema.py
@@ -198,7 +198,6 @@ def model_schema(
 
 
 def get_field_info_schema(field: ModelField, schema_overrides: bool = False) -> Tuple[Dict[str, Any], bool]:
-
     # If no title is explicitly set, we don't set title in the schema for enums.
     # The behaviour is the same as `BaseModel` reference, where the default title
     # is in the definitions part of the schema.

--- a/pydantic/types.py
+++ b/pydantic/types.py
@@ -481,6 +481,7 @@ else:
 
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ SET TYPES ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+
 # This types superclass should be Set[T], but cython chokes on that...
 class ConstrainedSet(set):  # type: ignore
     # Needed for pydantic to detect that this is a set
@@ -568,6 +569,7 @@ def confrozenset(
 
 
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ LIST TYPES ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
 
 # This types superclass should be List[T], but cython chokes on that...
 class ConstrainedList(list):  # type: ignore
@@ -1094,7 +1096,6 @@ class ByteSize(int):
 
     @classmethod
     def validate(cls, v: StrIntFloat) -> 'ByteSize':
-
         try:
             return cls(int(v))
         except ValueError:
@@ -1116,7 +1117,6 @@ class ByteSize(int):
         return cls(int(float(scalar) * unit_mult))
 
     def human_readable(self, decimal: bool = False) -> str:
-
         if decimal:
             divisor = 1000
             units = ['B', 'KB', 'MB', 'GB', 'TB', 'PB']
@@ -1135,7 +1135,6 @@ class ByteSize(int):
         return f'{num:0.1f}{final_unit}'
 
     def to(self, unit: str) -> float:
-
         try:
             unit_div = BYTE_SIZES[unit.lower()]
         except KeyError:

--- a/tests/test_dataclasses.py
+++ b/tests/test_dataclasses.py
@@ -5,6 +5,7 @@ import re
 import sys
 from collections.abc import Hashable
 from datetime import datetime
+from functools import cached_property
 from pathlib import Path
 from typing import Callable, ClassVar, Dict, FrozenSet, List, Optional, Set, Union
 
@@ -1631,3 +1632,31 @@ def test_can_deepcopy_wrapped_dataclass_type():
     B = copy.deepcopy(A)
     assert B is not A
     assert B(1) == A(1)
+
+def test_cached_property():
+    @dataclasses.dataclass(frozen=True)
+    class A:
+        name: str
+
+        @cached_property
+        def _name(self): 
+            return "name"
+
+    class MyModel(BaseModel, arbitrary_types_allowed=True, frozen=True, extra=Extra.forbid):
+        scheduler: A
+
+    models = {
+        "AX": MyModel(scheduler=A("a")),
+    }
+
+    sched = A("sched")
+
+    models_2 = {
+        "AY": models["AX"].copy(update=dict(scheduler=sched)), 
+    }
+
+    models |= models_2
+
+    models["AY"].scheduler._name
+    MyModel.parse_obj(models["AY"].dict())
+        

--- a/tests/test_dataclasses.py
+++ b/tests/test_dataclasses.py
@@ -5,7 +5,12 @@ import re
 import sys
 from collections.abc import Hashable
 from datetime import datetime
-from functools import cached_property
+
+try:
+    from functools import cached_property
+except ImportError:
+    pass
+
 from pathlib import Path
 from typing import Callable, ClassVar, Dict, FrozenSet, List, Optional, Set, Union
 
@@ -1633,30 +1638,31 @@ def test_can_deepcopy_wrapped_dataclass_type():
     assert B is not A
     assert B(1) == A(1)
 
+
+@pytest.mark.skipif(sys.version_info < (3, 8), reason='cached_property was introduced in python3.8 stdlib')
 def test_cached_property():
     @dataclasses.dataclass(frozen=True)
     class A:
         name: str
 
         @cached_property
-        def _name(self): 
-            return "name"
+        def _name(self):
+            return 'name'
 
     class MyModel(BaseModel, arbitrary_types_allowed=True, frozen=True, extra=Extra.forbid):
         scheduler: A
 
     models = {
-        "AX": MyModel(scheduler=A("a")),
+        'AX': MyModel(scheduler=A('a')),
     }
 
-    sched = A("sched")
+    sched = A('sched')
 
     models_2 = {
-        "AY": models["AX"].copy(update=dict(scheduler=sched)), 
+        'AY': models['AX'].copy(update=dict(scheduler=sched)),
     }
 
-    models |= models_2
+    models = {**models, **models_2}
 
-    models["AY"].scheduler._name
-    MyModel.parse_obj(models["AY"].dict())
-        
+    models['AY'].scheduler._name
+    MyModel.parse_obj(models['AY'].dict())

--- a/tests/test_generics.py
+++ b/tests/test_generics.py
@@ -745,7 +745,6 @@ def test_generic_model_from_function_pickle_fail(create_module):
 
 
 def test_generic_model_redefined_without_cache_fail(create_module, monkeypatch):
-
     # match identity checker otherwise we never get to the redefinition check
     monkeypatch.setattr('pydantic.generics.all_identical', lambda left, right: False)
 
@@ -975,7 +974,6 @@ def test_replace_types_with_user_defined_generic_type_field():
         pass
 
     class Model(GenericModel, Generic[T, KT, VT]):
-
         map_field: GenericMapping[KT, VT]
         list_field: GenericList[T]
 
@@ -1323,7 +1321,6 @@ def test_generic_with_user_defined_generic_field():
         pass
 
     class Model(GenericModel, Generic[T]):
-
         field: GenericList[T]
 
     model = Model[int](field=[5])

--- a/tests/test_main.py
+++ b/tests/test_main.py
@@ -2186,7 +2186,6 @@ def test_class_kwargs_config_json_encoders():
 
 
 def test_class_kwargs_config_and_attr_conflict():
-
     with pytest.raises(
         TypeError, match='Specifying config in two places is ambiguous, use either Config attribute or class kwargs'
     ):

--- a/tests/test_rich_repr.py
+++ b/tests/test_rich_repr.py
@@ -25,7 +25,6 @@ def test_rich_repr() -> None:
 
 
 def test_rich_repr_color() -> None:
-
     color = Color((10, 20, 30, 0.1))
     rich_repr = list(color.__rich_repr__())
 

--- a/tests/test_typing.py
+++ b/tests/test_typing.py
@@ -103,7 +103,6 @@ def test_convert_generics(type_, expectations):
 @pytest.mark.skipif(sys.version_info < (3, 10), reason='NewType class was added in python 3.10.')
 def test_convert_generics_unsettable_args():
     class User(NewType):
-
         __origin__ = type(list[str])
         __args__ = (list['Hero'],)
 

